File: trunk/esper/src/main/java/net/esper/view/ZeroDepthStream.java
/**************************************************************************************
 * Copyright (C) 2006 Esper Team. All rights reserved.                                *
 * http://esper.codehaus.org                                                          *
 * ---------------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the GPL license       *
 * a copy of which has been included with this distribution in the license.txt file.  *
 **************************************************************************************/
package net.esper.view;

import java.util.Iterator;
import java.util.List;
import java.util.LinkedList;

import net.esper.event.EventType;
import net.esper.event.EventBean;
import net.esper.collection.NullIterator;
import net.esper.collection.SingleEventIterator;
import net.esper.util.ExecutionPathDebugLog;
import org.apache.commons.logging.LogFactory;
import org.apache.commons.logging.Log;

/**
 * Event stream implementation that does not keep any window by itself of the events coming into the stream.
 */
public final class ZeroDepthStream implements EventStream
{
    private final LinkedList<View> children = new LinkedList<View>();
    private final EventType eventType;
    private EventBean lastInsertedEvent;

    /**
     * Ctor.
     * @param eventType - type of event
     */
    public ZeroDepthStream(EventType eventType)
    {
        this.eventType = eventType;
    }

    public final void insert(EventBean event)
    {
        if ((ExecutionPathDebugLog.isDebugEnabled) && (log.isDebugEnabled()))
        {
            log.debug(".insert Received event, updating child views, event=" + event);
        }

        // Get a new array created rather then re-use the old one since some client listeners
        // to this view may keep reference to the new data
        EventBean[] row = new EventBean[]{event};
        for (View childView : children)
        {
            childView.update(row, null);
        }

        lastInsertedEvent = event;
    }

    public final EventType getEventType()
    {
        return eventType;
    }

    public final Iterator<EventBean> iterator()
    {
        return new SingleEventIterator(lastInsertedEvent);
    }

    public final View addView(View view)
    {
        children.add(view);
        view.setParent(this);
        return view;
    }

    public final List<View> getViews()
    {
        return children;
    }

    public final boolean removeView(View view)
    {
        boolean isRemoved = children.remove(view);
        view.setParent(null);
        return isRemoved;
    }

    public final boolean hasViews()
    {
        return (!children.isEmpty());
    }

    private static final Log log = LogFactory.getLog(ZeroDepthStream.class);
}

File: NEsper/view/ZeroDepthStream.cs
using System;
using System.Collections.Generic;

using net.esper.compat;
using net.esper.events;

using LogFactory = org.apache.commons.logging.LogFactory;
using Log = org.apache.commons.logging.Log;

namespace net.esper.view
{
	/// <summary>
    /// Event stream implementation that does not keep any window by itself of the events
    /// coming into the stream.
    /// </summary>
    
    public sealed class ZeroDepthStream : EventStream
    {
        private readonly ELinkedList<View> children = new ELinkedList<View>();
        private readonly EventType eventType;

        /// <summary>
        /// Initializes a new instance of the <see cref="ZeroDepthStream"/> class.
        /// </summary>
        /// <param name="eventType">type of event</param>

        public ZeroDepthStream(EventType eventType)
        {
            this.eventType = eventType;
        }

        /// <summary>
        /// Inserts the specified event.
        /// </summary>
        /// <param name="ev">The ev.</param>
        
        public void Insert(EventBean ev)
        {
            if (log.IsDebugEnabled)
            {
                log.Debug(".insert Received event, updating child views, event=" + ev);
            }

			// Get a new array created rather then re-use the old one since some client listeners
			// to this view may keep reference to the new data
			EventBean[] row = new EventBean[]{ev};
			foreach (View childView in children)
			{
				childView.Update(row, null);
			}
        }

        /// <summary>
        /// Provides metadata information about the type of object the event collection contains.
        /// </summary>
        /// <value></value>
        /// <returns> metadata for the objects in the collection
        /// </returns>
        public EventType EventType
        {
            get { return eventType; }
        }

        /// <summary>
        /// Returns an enumerator that iterates through the collection.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.Collections.Generic.IEnumerator`1"></see> that can be used to iterate through the collection.
        /// </returns>
        public IEnumerator<EventBean> GetEnumerator()
        {
            return null;
        }

        /// <summary>
        /// Add a view to the viewable object.
        /// </summary>
        /// <param name="view">to add</param>
        /// <returns>view to add</returns>
        public View AddView(View view)
        {
            children.Add(view);
            view.Parent = this;
            return view;
        }

        /// <summary>
        /// Returns all added views.
        /// </summary>
        /// <returns>list of added views</returns>
        public IList<View> Views
        {
        	get { return children; }
        }

        /// <summary>
        /// Remove a view.
        /// </summary>
        /// <param name="view">to remove</param>
        /// <returns>
        /// true to indicate that the view to be removed existed within this view, false if the view to
        /// remove could not be found
        /// </returns>
        public Boolean RemoveView(View view)
        {
            Boolean isRemoved = children.Remove(view);
            view.Parent = null;
            return isRemoved;
        }

        /// <summary>
        /// Test is there are any views to the Viewable.
        /// </summary>
        /// <value></value>
        /// <returns> true indicating there are child views, false indicating there are no child views
        /// </returns>
        public Boolean HasViews
        {
            get { return (children.Count > 0); }
        }

        private static readonly Log log = LogFactory.GetLog(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        #region IEnumerable Members

        /// <summary>
        /// Returns an enumerator that iterates through a collection.
        /// </summary>
        /// <returns>
        /// An <see cref="T:System.Collections.IEnumerator"></see> object that can be used to iterate through the collection.
        /// </returns>
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        #endregion
    }
}
